vernac/vernacentries.ml

```diff
@@ -1069,31 +1068,53 @@ let add_subnames_of ns full_n n =
         let n_elim = Id.of_string (Id.to_string mip.mind_typename ^ s) in
         match Nametab.extended_global_of_path (Libnames.make_path dp n_elim) with
         | exception Not_found -> ns
-        | n_elim -> NSet.add n_elim ns)
+        | n_elim -> (len, n_elim) :: ns)
       ns Sorts.all_families
 
-let interp_filter_in m = function
-  | ImportAll ->  Libobject.Unfiltered
-  | ImportNames ns ->
-    let module NSet = Globnames.ExtRefSet in
-    let dp_m = Nametab.dirpath_of_module m in
-    let ns =
-      List.fold_left (fun ns (n,etc) ->
-          let full_n =
-            let dp_n,n = repr_qualid n in
-            make_path (append_dirpath dp_m dp_n) n
-          in
-          let n = try Nametab.extended_global_of_path full_n
-            with Not_found ->
-              CErrors.user_err
-                Pp.(str "Cannot find name " ++ pr_qualid n ++ spc() ++
-                    str "in module " ++ pr_qualid (Nametab.shortest_qualid_of_module m))
-          in
-          let ns = NSet.add n ns in
-          if etc then add_subnames_of ns full_n n else ns)
-        NSet.empty ns
-    in
-    Libobject.Names ns
+let interp_names m ns =
+  let dp_m = Nametab.dirpath_of_module m in
+  let ns =
+    List.fold_left (fun ns (n,etc) ->
+        let len, full_n =
+          let dp_n,n = repr_qualid n in
+          List.length (DirPath.repr dp_n), make_path (append_dirpath dp_m dp_n) n
+        in
+        let n = try Nametab.extended_global_of_path full_n
+          with Not_found ->
+            CErrors.user_err
+              Pp.(str "Cannot find name " ++ pr_qualid n ++ spc() ++
+                  str "in module " ++ pr_qualid (Nametab.shortest_qualid_of_module m))
+        in
+        (* TODO dumpglob? *)
+        let ns = (len,n) :: ns in
+        if etc then add_subnames_of len ns full_n n else ns)
+      [] ns
+  in
+  ns
+
+let cache_name (len,n) =
+  let open Globnames in
+  let open GlobRef in
+  match n with
+  | SynDef kn -> Syntax_def.import_syntax_constant (len+1) (Nametab.path_of_syndef kn) kn
+  | TrueGlobal (VarRef _) -> assert false
+  | TrueGlobal gr ->
+    (* imports even local constants, do we care? *)
+    Nametab.(push (Exactly (len+1)) (path_of_global gr) gr)
```

Comment:
With this code if we do `Import M(somelocalconstant)` the locality will be ignored.
Previously the import would be a noop.
Not sure which behaviour seems more correct.

Reply:
Even an error might be appropriate, since the intentions disagree...

Reply:
A warning would be better, that way you can turn it into an error if desired.

Reply:
@SkySkimmer ping. Personally I'd favour a noop + a warning that it's actually a noop.

Reply:
:+1: for the noop+warning: noop keeps today’s semantics, and the warning can be turned into an error (as said) or silenced. More power to users!

Reply:
done